### ISSUE TITLE
Fatal Exception: java.lang.NullPointerException

### ISSUE THREAD

**Issue by tecbot** [adjust/flutter_sdk#30]
Since the integration of V2 Android Plugin (https://github.com/adjust/flutter_sdk/commit/efa0692d5832f73fd7e7f09ef8557c56ad7facad)  we get some crashes in our Android App:

```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
       at com.adjust.sdk.flutter.AdjustSdk$1.onAttributionChanged(AdjustSdk.java:10)
       at com.adjust.sdk.ActivityHandler$40.run(ActivityHandler.java:3)
       at android.os.Handler.handleCallback(Handler.java:883)
       at android.os.Handler.dispatchMessage(Handler.java:100)
       at android.os.Looper.loop(Looper.java:237)
       at android.app.ActivityThread.main(ActivityThread.java:8167)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
```

**Comment by oke331** [comment]
I get alomost simular crashes  in my Android app when I use PlatformView and startactivity in kotlin code.

```
java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
 	at com.adjust.sdk.flutter.AdjustSdk$4.onFinishedEventTrackingSucceeded(AdjustSdk.java:463)
 	at com.adjust.sdk.ActivityHandler$36.run(ActivityHandler.java:1245)
 	at android.os.Handler.handleCallback(Handler.java:938)
 	at android.os.Handler.dispatchMessage(Handler.java:99)
 	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by Fatimamostafa** [comment]
> I get alomost simular crashes in my Android app when I use MethodChannel and startactivity in kotlin code.
> 
> ```
> java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
>  	at com.adjust.sdk.flutter.AdjustSdk$4.onFinishedEventTrackingSucceeded(AdjustSdk.java:463)
>  	at com.adjust.sdk.ActivityHandler$36.run(ActivityHandler.java:1245)
>  	at android.os.Handler.handleCallback(Handler.java:938)
>  	at android.os.Handler.dispatchMessage(Handler.java:99)
>  	at android.os.Looper.loop(Looper.java:223)
> 	at android.app.ActivityThread.main(ActivityThread.java:7660)
> 	at java.lang.reflect.Method.invoke(Native Method)
> 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
>  	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
> ```

I'm having this same issue. Did you figure a fix?

**Comment by uerceg** [comment]
Hey guys.

We have checked this issue and managed to reproduce it (thanks to @Fatimamostafa for reproduction steps in linked issue).

So issue seems to be the fact that **MethodChannel** used to communicate between native and Flutter layer gets nulled natively in Android in scenario in which app is exited with back gesture/button. Upon exiting the app in this way, plugin gets [onDetachedFromEngine](https://github.com/adjust/flutter_sdk/blob/master/android/src/main/java/com/adjust/sdk/flutter/AdjustSdk.java#L88) method pinged (as expected) and upon returning into the app **onAttachedToEngine** gets pinged (as it should) and in there **MethodChannel** gets [recreated](https://github.com/adjust/flutter_sdk/blob/master/android/src/main/java/com/adjust/sdk/flutter/AdjustSdk.java#L83).

However, in moment in which our native SDK sends something back to callback we define natively in **AdjustSdk.java** of our Flutter SDK (for example [event tracking success](https://github.com/adjust/flutter_sdk/blob/master/android/src/main/java/com/adjust/sdk/flutter/AdjustSdk.java#L453) callback gets pinged), **channel** (for now) mysteriously becomes **null** causing this exception. If app is backgrounded and brought back to foreground and/or killed and restarted, this doesn't happen.

For now there are few ways which will prevent crash from happening of course, but none of them actually provides solution which will also allow callbacks to be pinged in Dart. Two ways how it might be solved:

1. Simply check if **channel** is **null** before trying to fire callback and crash will not happen. But this means that we're not gonna attempt to ping any callback which you have defined in Dart.
2. We can try to [skip destroying method channel](https://github.com/adjust/flutter_sdk/blob/master/android/src/main/java/com/adjust/sdk/flutter/AdjustSdk.java#L91-L94) when detaching from engine and in this scenario upon starting the app in problematic scenario, **channel** is actually not **null** when we want to ping callbacks in Dart. **However** when this is attempted, it fail (without crashing) and reports following error message:
```
W/FlutterJNI: Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: com.adjust.sdk/api. Response ID: 0
```

We're still working on this, but that's the current state of things. Right now we can fix this to prevent crashes and that will definitely be part of next SDK update, but we would also like to see if there's a way to make Dart callbacks to be pinged in this scenario as well.

In case this situation rings a bell to anyone, any advice is more than welcomed. Nevertheless, will keep you posted on update progress.

Cheers

**Comment by Fatimamostafa** [comment]
I've figured a way to transcend the crash yet get Dart callbacks to be pinged. 

`    override fun onDestroy() {
        flutterEngine?.platformViewsController?.detach()
        super.onDestroy()
    }
`

However, this will still cause a crash when it tries to call super.onDestroy(). But doesn't affect the user in terms of any experience or Adjust callbacks.

**Comment by uerceg** [comment]
Hey guys.

We have just released [v4.26.0](https://github.com/adjust/flutter_sdk/releases/tag/v4.26.0) update where NPEs should hopefully be addressed, but solution is not really perfect. We have opened [issue in Flutter repo](https://github.com/flutter/flutter/issues/76526) where you can find all details and hopefully we can get some assistance from Flutter maintainers on how to handle this case. From what we saw so far, nothing from available solutions we managed to didn't help to solve issue completely. But at least, NPEs should be gone.

As soon as we get some update in above mentioned ticket and hopefully suggested solution, we're gonna add it to our SDK and release an update ASAP.

I'll close this issue for now, but in case you still notice NPEs happening even with this (or later) versions, feel free to reopen / comment. I'm also gonna continue to post updates in here regarding Flutter repo ticket.

Thank you all for reporting this issue and with all the help.

Cheers
